source/common/common/key_value_store_base.cc

```diff
@@ -0,0 +1,102 @@
+#include "source/common/common/key_value_store_base.h"
+
+namespace Envoy {
+namespace {
+
+// Removes a length prefixed token from |contents| and returns the token,
+// or returns absl::nullopt on failure.
+absl::optional<absl::string_view> getToken(absl::string_view& contents) {
+  const auto it = contents.find("\n");
+  if (it == contents.npos) {
+    ENVOY_LOG_MISC(warn, "Bad file: no newline");
+    return {};
+  }
+  uint64_t length;
+  if (!absl::SimpleAtoi(contents.substr(0, it), &length)) {
+    ENVOY_LOG_MISC(warn, "Bad file: no length");
+    return {};
+  }
+  contents.remove_prefix(it + 1);
+  if (contents.size() < length) {
+    ENVOY_LOG_MISC(warn, "Bad file: insufficient contents");
+    return {};
+  }
+  absl::string_view token = contents.substr(0, length);
+  contents.remove_prefix(length);
+  return token;
+}
+
+bool tokenizeContents(absl::string_view contents,
+                      absl::flat_hash_map<std::string, std::string>& store) {
+  // Assuming |contents| is in the format
+  // [length]\n[key]\n[length]\n[value]
+  // tokenizes contents into the provided store.
+  // This is best effort, and will return false on failure without clearing
+  // partially parsed data.
+  while (!contents.empty()) {
+    absl::optional<absl::string_view> key = getToken(contents);
+    absl::optional<absl::string_view> value = getToken(contents);
+    if (!key.has_value() || !value.has_value()) {
+      return false;
+    }
+    store.emplace(std::string(key.value()), std::string(value.value()));
+  }
+  return true;
+}
+
+} // namespace
+
+KeyValueStoreBase::KeyValueStoreBase(Event::Dispatcher& dispatcher,
+                                     std::chrono::seconds flush_interval)
+    : flush_timer_(dispatcher.createTimer([this]() { flush(); })) {
+  flush_timer_->enableTimer(flush_interval);
+}
+
+void KeyValueStoreBase::addOrUpdateKey(absl::string_view key, absl::string_view value) {
+  store_.erase(key);
+  store_.emplace(key, value);
+}
+
+void KeyValueStoreBase::removeKey(absl::string_view key) { store_.erase(key); }
+absl::string_view KeyValueStoreBase::getKey(absl::string_view key) {
+  auto it = store_.find(key);
+  if (it == store_.end()) {
+    return "";
```

Comment:
Do we want this to return optional to distinguish empty string from unset?

Reply:
I couldn't think of a use case but I'd waffled on it too.  done.